### PR TITLE
Retry for adding remote cluster

### LINKED ISSUE #38988: CDS and EDS will be partially lost when the network connection from istiod to  remote APIServer is abnormal
### Bug Description

We recreate the `istiod` pods when the network connection from `istiod` to  remote `APIServer` is abnormal, two minutes later, we let the network connection returned to normal, the CDS and EDS of ingressgateway will be partially lost.

- Before reconnecting from `istiod` to remote `APIServer`, the `CDS`of ingressgateway is as follows:
```
$ istioctl pc c istio-ingressgateway-5d5b776b7b-pq76c.istio-system
SERVICE FQDN                                            PORT      SUBSET     DIRECTION     TYPE           DESTINATION RULE
BlackHoleCluster                                        -         -          -             STATIC
agent                                                   -         -          -             STATIC
details.default.svc.cluster.local                       9080      -          outbound      EDS
hpa-metrics-service.kube-system.svc.cluster.local       443       -          outbound      EDS
istio-crd.istio-system.svc.cluster.local                61011     -          outbound      EDS
istio-ingressgateway.istio-system.svc.cluster.local     80        -          outbound      EDS
istiod-1-12-6.istio-system.svc.cluster.local            443       -          outbound      EDS
istiod-1-12-6.istio-system.svc.cluster.local            15012     -          outbound      EDS
istiod-1-12-6.istio-system.svc.cluster.local            15014     -          outbound      EDS
kube-dns.kube-system.svc.cluster.local                  53        -          outbound      EDS
kubernetes.default.svc.cluster.local                    443       -          outbound      EDS
metrics-service.kube-system.svc.cluster.local           443       -          outbound      EDS
productpage.default.svc.cluster.local                   9080      -          outbound      EDS
prometheus_stats                                        -         -          -             STATIC
ratings.default.svc.cluster.local                       9080      -          outbound      EDS
reviews.default.svc.cluster.local                       9080      -          outbound      EDS
sds-grpc                                                -         -          -             STATIC
xds-grpc                                                -         -          -             STATIC
zipkin                                                  -         -          -             STRICT_DNS
zipkin.istio-system.svc.cluster.local                   9411      -          outbound      EDS
```

- After reconnecting from `istiod` to remote `APIServer`, the `CDS` of ingressgatewayis as follows:
```
istioctl pc c istio-ingressgateway-5d5b776b7b-pq76c.istio-system
SERVICE FQDN                             PORT     SUBSET     DIRECTION     TYPE           DESTINATION RULE
BlackHoleCluster                         -        -          -             STATIC
agent                                    -        -          -             STATIC
kubernetes.default.svc.cluster.local     443      -          outbound      EDS
prometheus_stats                         -        -          -             STATIC
sds-grpc                                 -        -          -             STATIC
xds-grpc                                 -        -          -             STATIC
zipkin                                   -        -          -             STRICT_DNS
```

### Version

```prose
$ istioctl version                                                                
client version: 1.9.4
control plane version: 1.12.6
data plane version: 1.12.6 (7 proxies)
$ kubectl version --short
Client Version: v1.24.0
Kustomize Version: v4.5.4
Server Version: v1.20.6-tke.17
```

### Additional Information

$ istioctl bug-report

Target cluster context: cls

Running with the following config: 

istio-namespace: istio-system
full-secrets: false
timeout (mins): 30
include: {  }
exclude: { Namespaces: kube-system, kube-public, kube-node-lease, local-path-storage } AND { Namespaces: kube-system, kube-public, kube-node-lease, local-path-storage }
end-time: 2022-05-18 10:24:21.069784 +0800 CST

The following Istio control plane revisions/versions were found in the cluster:
Revision 1-12-6:
no running Istio pods in "istio-system"

The following proxy revisions/versions were found in the cluster:

Fetching proxy logs for the following containers:

default/details-v1/details-v1-c5d4c46c5-6jvpc/details
default/details-v1/details-v1-c5d4c46c5-6jvpc/istio-proxy
default/productpage-v1/productpage-v1-558d748d94-8wcbc/istio-proxy
default/productpage-v1/productpage-v1-558d748d94-8wcbc/productpage
default/ratings-v1/ratings-v1-5cd5486d45-vgx7c/istio-proxy
default/ratings-v1/ratings-v1-5cd5486d45-vgx7c/ratings
default/reviews-v1/reviews-v1-566bcb47c4-xc5qc/istio-proxy
default/reviews-v1/reviews-v1-566bcb47c4-xc5qc/reviews
default/reviews-v2/reviews-v2-848dd747c5-d6cjp/istio-proxy
default/reviews-v2/reviews-v2-848dd747c5-d6cjp/reviews
default/reviews-v3/reviews-v3-5947f7b997-r7c2l/istio-proxy
default/reviews-v3/reviews-v3-5947f7b997-r7c2l/reviews
istio-system/istio-ingressgateway/istio-ingressgateway-5d5b776b7b-pq76c/istio-proxy
prom-0u3pw1a1/proxy-agent/proxy-agent-8597d5fff6-krjbm/proxy-agent

Fetching Istio control plane information from cluster.

The following fetches are still running: 
  kubectl get --all-namespaces logconfigs.cls.cloud.tencent.com,podmonitors.monitoring.coreos.com,probes.monitoring.coreos.com,prometheuses.monitoring.coreos.com,prometheusrules.monitoring.coreos.com,servicemonitors.monitoring.coreos.com,tkeserviceconfigs.cloud.tencent.com,volumesnapshotclasses.snapshot.storage.k8s.io,volumesnapshotcontents.snapshot.storage.k8s.io,volumesnapshots.snapshot.storage.k8s.io -o yaml

Running istio analyze on namespace default.
Running istio analyze on namespace istio-system.

### PR DESCRIPTION
**Pull request by miaoyq**
**Please provide a description of this PR:**

Fix #38988
/cc @hzxuzhonghu @howardjohn

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func newEndpointSliceController(c *Controller) *endpointSliceController {
 
 // TODO use this to automatically switch to EndpointSlice mode
 func endpointSliceV1Available(client kubelib.Client) bool {
-	return client != nil && kubelib.IsAtLeastVersion(client, 21)
+	return client != nil && !kubelib.IsLessThanVersion(client, 21)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointslice.go:83]: Why is this changed?

File: pkg/kube/multicluster/secretcontroller.go [hunk 2]
```diff
@@ -196,7 +199,9 @@ func (c *Controller) processItem(key types.NamespacedName) error {
 	}
 	if exists {
 		log.Debugf("secret %s exists in informer cache, processing it", key)
-		c.addSecret(key, obj.(*corev1.Secret))
+		if err := c.addSecret(key, obj.(*corev1.Secret)); err != nil {
+			return fmt.Errorf("error adding secret %s error: %v", key, err)
```
Comment (howardjohn) [inline pkg/kube/multicluster/secretcontroller.go:203]:
```suggestion
			return fmt.Errorf("error adding secret %s: %v", key, err)
```

nit: go error common practice  is to just add `: err`

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go (hunk 1 above)
Comment (miaoyq) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:83]: We should use `EndpointOnly` when getting the version fails, otherwise clusters less than vertion 21 will try to watch endpointslice(v1).

File: pkg/kube/multicluster/secretcontroller.go (hunk 2 above)
Comment (miaoyq) [reply pkg/kube/multicluster/secretcontroller.go:203]: Thx, will fix  :-)

File: pkg/kube/multicluster/secretcontroller.go [hunk 3]
```diff
@@ -116,7 +117,9 @@ func NewController(kubeclientset kube.Client, namespace string, localClusterID c
 		cs:                 newClustersStore(),
 		informer:           secretsInformer,
 	}
-	controller.queue = controllers.NewQueue("multicluster secret", controllers.WithReconciler(controller.processItem))
+	controller.queue = controllers.NewQueue("multicluster secret",
+		controllers.WithMaxAttempts(maxRetries),
```
[the remark this answers is not in the document]
Comment (miaoyq) [reply pkg/kube/multicluster/secretcontroller.go:121]: Thx  :-)

**Review by hzxuzhonghu — COMMENTED** [review]
TestBadRemoteSecret failure is related

**Comment by hzxuzhonghu** [comment]
The test failure caused by istiod not ready because of retry

**Comment by miaoyq** [comment]
@hzxuzhonghu Thank you for reminding.

**Comment by miaoyq** [comment]
@hzxuzhonghu Thx for reminding.  But in my test environment, using the new rateLimiter, `istiod` startup only takes `30s`, retry does not block `istiod` startup.

**Comment by stevenctl** [comment]
I think the issue is that `HasSynced` _must_ for secretcontroller _must_ process all the secrets present at startup. Do we want to just apply the "sync timeout" there as well?

**Comment by miaoyq** [comment]
@stevenctl When adding a cluster fails or timeout, we need remove the cluster from remoteClusters immediately so that it does not block istiod startup.

**Comment by qudongfang** [comment]
Are you guys still working on this PR? It seems pretty useful to me.

**Comment by hzxuzhonghu** [comment]
@miaoyq Do you have bandwidth to continue with this?

**Comment by Minh-Ng** [comment]
@hzxuzhonghu ran into this in https://github.com/istio/istio/issues/38988 and https://github.com/istio/istio/issues/45992, if we exceed maxRetries is it true that we will never retry to discover the remote cluster?
